### ISSUE TITLE
Update to 0.5.1

### ISSUE THREAD

**Issue by Kekun** [libretro/blastem-libretro#4]
The current version of BlastEm used here is a bit after 0.3.1, at the time the emulation was quite innacurate. The latest version is 0.5.1 and tons of improvements happened, BlastEm is now a solid Mega Drive and Master System emulator.

It would be great to see this core updated. I spent quite some hours and try various ways to update it, with no success. Not managing to fix that myself, I to resort to ask for an update from someone more used to this kind of work than I am.

**Comment by r-type** [comment]
preliminary update of core to version blastem-f2ee46d08b01 0.5.2-pre

https://github.com/r-type/blastem-libretro

![blastemhg](https://user-images.githubusercontent.com/3614417/30787734-d48beca2-a18f-11e7-8af2-7432db426dc0.png)

this is a rough update,  still has to do the finishes.
but video/sound/input at least works for genesis.

**Comment by Kekun** [comment]
@r-type I just tried your branch in GNOME Games, when selecting a Mega Drive game the application crashes after printing this:
(gnome-games:15): BlastEm-CRITICAL **: Fatal Error: Failed to load ROM DB

The current version works fine. For some reason there were no backtrace. It may be caused by a bug in Games or a non-implemented Libretro feature, but well, here is what I got so far. :)

**Comment by r-type** [comment]
Just put rom.db in the current folfer.

That something that need to be handle properly in a later commit. Keep in
mind it is wip and at early stage

**Comment by r-type** [comment]
rom.db should now be loading from RA system directory.

**Comment by Kekun** [comment]
@r-type This looks like the kind of data we want like to compile and ship with the binary, and then access as a file to avoid tons of changes.

**Comment by r-type** [comment]
yeah , for now i only re-implement what was done before.
just fix it if rom.db is not found , it should load and not crashing now.

**Comment by Kekun** [comment]
Any news regarding this port? :)

**Comment by RobLoach** [comment]
To help in updating Blastem, I made a git mirror of the hg repo: https://github.com/robloach/blastem

In order to use it we would need to:

1. Move the repo over to https://github.com/libretro-mirrors
2. Rebase the repo on top of it

Tested out [r-type's blastem fork](https://github.com/r-type/blastem-libretro) and the update seems to work well! [Added rom.db to libretro-database](https://github.com/libretro/libretro-database/pull/547).

**Comment by Kekun** [comment]
BTW I asked Mike Pavone if he is be fine having the Libretro port upstream, and he is! Having it upstream would ease maintaining the Libretro API a lot.

**Comment by twinaphex** [comment]
I agree. I think he might have to re-port it though, probably a lot of code changed in the interim.

**Comment by gingerbeardman** [comment]
Blastem is currently at 0.6.0-pre

**Comment by retrorepair** [comment]
Bit disappointed this stalled, would love to see it in retroarch

**Comment by RobLoach** [comment]
## 0.6.1
Released January 4th, 2019

Fixed build script so controller images are actually included so UI doesn't crash
Disabled most bindings when UI active (fixes crashes/wonkiness when pressing certain keys)
Fixed Windows implementation of get_config_dir() so config file ends up in %localappdata%\blastem like it should
Fixed the location of sticky_path on all platforms
Added virtual root directory used by ROM UI to new Nuklear UI

## 0.6.0
Released December 31st, 2018

Download for 32-bit Linux
Download for 64-bit Linux
Download for Windows
Download for OSX
Download Source
New Features
New Nuklear UI with almost complete access to configuration options
Added support for creating SDL2 mappings from inside the emulator
Loading ROMs from ZIP archives is now supported
Loading gzip compressed ROMs is now supported
Internal screenshots can now be in PNG format
New VDP plane debug view
VDP debug views are now in separate windows
Experimental support for sync to video (not enabled by default)
Preliminary support for MegaWifi cart emulation
Bugfixes
Fixed a number of 68K decoder bugs (mostly illegal instruction decoding)
Fixed implementation of the UNLK instruction when register is a7
Fixed a number of memory leaks
Accuracy/Completeness Improvements
Added support for J-Cart emulation
Implemented Z80 interrupt mode 2
Fixed the timing of a number of 68K instructions
Fixed the timing of the Z80 JP cc, nn instruction
Fixed the order bytes of a word are written to VRAM from the FIFO (fixes minor corruption in Road Rash 3 bike menu)
Other Changes
Added support for Open GL ES in addition to the existing desktop GL support
Some small optimizations
Added ROM DB entry for Squirrel King to support it's copy protection
Added support for float32 audio output (fixes an issue with defautl SDL2 driver in Windows when using more recent SDL2 versions)

**Comment by mikepavone** [comment]
FYI: I have started on adding libretro support in the upstream blastem repo. It still needs work (properly dealing with H32 mode, interlacing, rom.db, etc.), but the basic stuff is working okay.

**Comment by retrorepair** [comment]
This is excellent news! I think along with CRT switching this should be about the most accurate emulated representation of the megadrive/genesis! Can't wait to try it out

**Comment by gingerbeardman** [comment]
Thanks @mikepavone !

**Comment by Awakened0** [comment]
Is upstream libretro support mature enough now for someone to add it to the buildbot for nightlies?

**Comment by gingerbeardman** [comment]
0.6.2 is the latest

**Comment by BlockABoots** [comment]
Yes it some one could port this over to Retroarch core would be much appreciated

**Comment by Awakened0** [comment]
It's possible to compile the core from upstream's source. It currently looks for rom.db in root (next to retroarch.exe). But there is an audio error when you try to run it. I mentioned it to @mikepavone already on the official Blastem Discord.

**Comment by RobLoach** [comment]
That's great news! Nicely done, all! https://www.retrodev.com/repos/blastem/file/tip

We'll likely want to update libretro-super to pull those sources instead of these. Unsure if buildbot supports mercurial, so thta's something we'll have to fix up.

**Comment by BlockABoots** [comment]
Sadly i wouldn't have a clue on how to compile a Windows build, i wasn't even aware there was a cycle accurate Megadrive emulator until yesterday thats why it perked my interest in if there was a RetroArch core of it

**Comment by r-type** [comment]
Nice! i have tested on linux last commit that fix libretro build form @mikepavone , everything works fine !

**Comment by retrorepair** [comment]
@r-type how would I got about building on linux for libretro?

**Comment by fr500** [comment]
make libblastem.so

**Comment by BlockABoots** [comment]
any plans on a windows release?

**Comment by fr500** [comment]
patience

**Comment by retrorepair** [comment]
Make libblastem.so

Gives me:

Make: *** no rule to make target 'libblastem.so'. Stop.

**Comment by gingerbeardman** [comment]
Doesn't BlastEm have it's own menu/settings cart?

**Comment by retrorepair** [comment]
Can't see that working with Retroarch, could be wrong?

**Comment by retrorepair** [comment]
Ok so having tested I can say it's pretty amazing! Even with default options.

I've made some slight changes to the source (a couple temporarily) and was hoping you could take a look @mikepavone 

I've forked the mercurial repo here: http://github.com/retrorepair/blastem

I've only changed a few things in libblastem.c, added accurate video and audio timings depending on region and temporarily changed the heights from 243+294 to 240+288 as the current heights break crt switching in RA. I understand this is the entire resolution but part of that res is vblank which we have to use for crt switching. I've not modified anything else so vertical position is now off, as would be expected.

**Comment by fr500** [comment]
You should give the bounty to the original dev. He fixed everything.
I added manual builds to the buildbot already.

I'll be working on buildbot integration eventually.

**Comment by fr500** [comment]
So to update this repo do:

```
git remote add upstream https://github.com/fr500/blastem.git
git fetch upstream
git reset --hard upstream/master
```

I'll keep my fork up to date with mercurial at least.

**Comment by RobLoach** [comment]
Brought the hg upstream to https://github.com/libretro-mirrors/blastem too. Added you on there, @fr500. Thanks!

**Comment by fr500** [comment]
you should reset this one to that, this repo has no purpose anymore

**Comment by BlockABoots** [comment]
Apparently the Windows build has 1 frame of lag more than the Genesis Plus GX Core, can this be reduced?

**Comment by fr500** [comment]
Can this issue be closed now? it serves no purpose
